lib/drivers/docker/index.js

```diff
@@ -0,0 +1,294 @@
+'use strict';
+
+var _ = require('lodash');
+var async = require('async');
+var bl = require('bl');
+var debug = require('debug')('strong-pm:docker:driver');
+var Docker = require('dockerode');
+var EventEmitter = require('events').EventEmitter;
+var Image = require('./image');
+var inherits = require('util').inherits;
+var isError = require('util').isError;
+
+module.exports = exports = DockerDriver;
+
+function DockerDriver(opts) {
+  if (!opts.baseDir || !opts.console || !opts.server) {
+    throw Error('Requires baseDir, console, and server');
+  }
+  if (!(this instanceof DockerDriver)) {
+    return new DockerDriver(opts);
+  }
+  EventEmitter.call(this);
+  this.docker = new Docker();
+
+  this.baseDir = opts.baseDir;
+  this.console = opts.console;
+  this.server = opts.server;
+
+  this.instances = {};
+
+  this.defaultStartOptions = {
+    profile: true,
+    trace: false,
+    size: 'STRONGLOOP_CLUSTER' in process.env ?
+      process.env.STRONGLOOP_CLUSTER : 'CPU',
+  };
+
+  this.on('request', this._onRequest.bind(this));
+}
+
+inherits(DockerDriver, EventEmitter);
+
+DockerDriver.prototype.setStartOptions = function(id, opts) {
+  debug('DockerDriver.setStartOptions(%j, %j)', id, opts);
+  // console.trace();
+  var instance = this._instance(id);
+  var original = _.clone(instance.startOpts);
+
+  instance.startOpts = _.merge(instance.startOpts, opts);
+
+  if (opts.size != null && opts.size !== original.size) {
+    debug('cluster size changed, sending set-size: %j => %j', id, opts);
+    this.requestOfInstance(id, {cmd: 'set-size', size: opts.size});
+  } else {
+    // debug('cluster size unchanged');
+  }
+};
+
+DockerDriver.prototype.removeInstance = function(id, cb) {
+  debug('DockerDriver.removeInstance(%j)', id);
+  var instance = this._instance(id);
+  if (instance.current) {
+    instance.current.kill('SIGKILL', cb);
+    delete instance.current;
+  } else {
+    setImmediate(cb);
+  }
+};
+
+DockerDriver.prototype.deployInstance = function(id, req, res) {
+  debug('DockerDriver.deployInstance(%j)', id);
+  var self = this;
+  var instance = this._containerById(id);
+  var image = Image.from(this, id, this.baseDir, req, res);
+  instance.next = image;
+  image.on('error', function(err) {
+    console.error('error deploying: %s, %j', err, err);
```

Comment:
is this debug code? This will kill pm, I've been using assert.ifError() to clearly indicate situations that cannot be dealt with, but it seems like deploying could fail, and if so, this would be bad.

Of, if an error event is expected to never occur, that's OK.

Possible deployment errors should be written back to the client using the `res` object.